[quant] Dequantize apot tensor

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79845
* #79940
* __->__ #79530
* #79499
* #79936

### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

asl3 committed:
[quant] Dequantize apot tensor

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"




[ghstack-poisoned]

asl3 commented on mypy.ini:
```diff
@@ -58,6 +58,9 @@ ignore_missing_imports = True
 [mypy-torch.for_onnx.onnx]
 ignore_missing_imports = True
 
+[mypy-torch.ao.quantization.experimental.observer]
```
Is it okay for me to modify mypy.ini? This rule just applies to my oberver.py file

HDCharles replied:
I think this file should be excluded from being tracked by git, its fine to modify it, it shouldn't be reflected here.

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang reviewed (changes requested):
let's first take a look at the quantize PR first before this one (see my comments here for reasons why)

asl3 commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,14 +1,41 @@
 
-    def dequantize(self) -> Tensor:
-        raise NotImplementedError
+    """ Dequantizes integer Tensor to floating point representation
```
agreed, i will rebase and add functionality to take in reduced precision fp representation

jerryzh168 commented on torch/ao/quantization/experimental/APoT_tensor.py:
```diff
@@ -1,14 +1,41 @@
+        max_val = 1.0
+
+        # make observer
+        obs = APoTObserver(max_val=max_val, b=b, k=k)
```
shouldn't dequantize take a quantized tensor as input? and should we store quantization parameters in the quantized tensor instead of using observer to get them?

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang reviewed (changes requested):
some of my comments in https://github.com/pytorch/pytorch/pull/79499 apply here as well, so let's address that first

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

dzdang commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -10,8 +10,8 @@
-        and hardcoded values for b, k by comparing to uniform observer
-        (non-uniform observer reduces to uniform for k = 1)
+        and hardcoded values for b, k by comparing to uniform quantization
+        (non-uniform quantization reduces to uniform for k = 1)
```
seems this was changed in the wrong PR, should be done in the previous PR? same goes for some of the other changes in this PR

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_quantized_tensor.py:
```diff
@@ -105,10 +105,110 @@ def test_quantize_APoT_reduced_precision(self):
+        # initialize quantize APoT tensor to dequantize:
+        # generate tensor with random values between 0 -> 2**4 = 16
+        # because there are 2**b = 2**4 quantization levels total
+        tensor2dequantize = 16 * torch.rand(size)
```
maybe rename this to `int_repr_for_qtensor`?

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantized_tensor.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 commented on test/quantization/core/experimental/test_quantizer.py:
```diff
@@ -77,9 +77,59 @@ def test_quantize_APoT_k2(self):
+
+        dequantized_result = quantizer.dequantize()
+
+        print(dequantized_result)
```
will delete!

asl3 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -23,6 +23,7 @@ def __init__(
         signed,
             dtype=torch.quint8) -> None:
         self.signed = signed
+        self.use_int_repr = True
```
fixed

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 commented on test/quantization/core/experimental/test_quantizer.py:
```diff
@@ -77,9 +77,59 @@ def test_quantize_APoT_k2(self):
+
+        quantized_result = quantizer.quantize_APoT(tensor2quantize=dequantized_result)
+
+        self.assertTrue(torch.equal(dequantized_result, quantized_result))
```
fixed, should be comparing equality between quantized_result and orig_input

dzdang replied:
torch.Equal is still in the new version. why is this needed? instead of just using assertTrue?

asl3 replied:
@dzdang i am asserting that quantized_result equals orig_input, and i need torch.equal to do that (assertEqual wouldn't work for tensors)

dzdang replied:
we use assertEqual a lot to compare tensors. what is the issue you're having?

asl3 replied:
Using `assertEqual(tensor1, tensor2)` gives me the error: `... in _baseAssertEqual if not first == second: RuntimeError: Boolean value of Tensor with more than one value is ambiguous` , seems like `assertTrue(torch.equal(tensor1, tensor2)` is a work-around

dzdang replied:
if the test class inherits from torch.testing._internal.common_utils.TestCase, it shoudl work

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -54,8 +54,24 @@ def quantize_APoT(self, tensor2quantize: Tensor):
+        result: floating point representation of input Tensor
+    """
+    def dequantize(self):  # type: ignore[override]
+        tensor2dequantize = self.data.float()
```
nit: this name is a little weird to me since tensor more or less just means matrix, I would maybe consider something more descriptive.

HDCharles replied:
i.e. float2apot or float2quantized maybe

jerryzh168 commented on test/quantization/core/experimental/test_quantizer.py:
```diff
@@ -72,14 +72,63 @@ def test_quantize_APoT_k2(self):
+        # because there are 2**b = 2**4 quantization levels total
+        tensor2dequantize = 16 * torch.rand(size)
+        quantizer = APoTQuantizer(4, 2, 1.0, False)
+        quantizer.data = tensor2dequantize.int()
```
why `data` is stored in quantizer? I feel we should store it in a quantized tensor

HDCharles reviewed (approved):
looks good, if you and david have discussed the name already feel free to just ignore

asl3 committed:
Update on "[quant] Dequantize apot tensor"


### Summary:
This PR implements functionality to dequantize input APoT quantized tensors. The dequantize method overloads the inherited dequantize method from the parent class: https://pytorch.org/docs/stable/generated/torch.dequantize.html. The method takes an integer tensor as input and converts to floating point representation based on the calculated quantization levels from the APoT non-uniform observer. This approach follows the methodology outlined in the paper https://arxiv.org/pdf/1909.13144.pdf.

### Test Plan:
Run unit tests with: `python pytorch/test/quantization/core/experimental/test_quantizer.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
